src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs

```diff
@@ -0,0 +1,161 @@
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (projectSnapshotManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManager));
+            }
+
+            _dotNetProject = project;
+            _foregroundDispatcher = foregroundDispatcher;
+            _projectSnapshotManager = projectSnapshotManager;
+            _projectChangedSemaphore = new AsyncSemaphore(initialCount: 1);
+
+            AttachToProject();
+        }
+
+        public DotNetProject DotNetProject => _dotNetProject;
+
+        public HostProject HostProject => _currentHostProject;
+
+        protected ForegroundDispatcher ForegroundDispatcher => _foregroundDispatcher;
+
+        public void Detatch()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified -= DotNetProject_Modified;
+
+            UpdateHostProjectForeground(null);
+        }
+
+        protected abstract Task OnProjectChangedAsync();
+
+        // Protected virtual for testing
+        protected virtual void AttachToProject()
+        {
+            ForegroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified += DotNetProject_Modified;
+
+            // Trigger the initial update to the project.
+            _batchingProjectChanges = true;
+            Task.Factory.StartNew(ProjectChangedBackgroundAsync, null, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.BackgroundScheduler);
+        }
+
+        // Must be called inside the lock.
+        protected async Task UpdateHostProjectUnsafeAsync(HostProject newHostProject)
+        {
+            _foregroundDispatcher.AssertBackgroundThread();
+
+            await Task.Factory.StartNew(UpdateHostProjectForeground, newHostProject, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.ForegroundScheduler);
+        }
+
+        protected async Task ExecuteWithLockAsync(Func<Task> func)
+        {
+            using (await _projectChangedSemaphore.EnterAsync().ConfigureAwait(false))
+            {
+                await func().ConfigureAwait(false);
+            }
+        }
+
+        private async Task ProjectChangedBackgroundAsync(object state)
+        {
+            ForegroundDispatcher.AssertBackgroundThread();
+
+            _batchingProjectChanges = false;
+
+            await OnProjectChangedAsync();
+        }
+
+        private void DotNetProject_Modified(object sender, SolutionItemModifiedEventArgs args)
+        {
+            if (args == null)
+            {
+                throw new ArgumentNullException(nameof(args));
+            }
+
+            ForegroundDispatcher.AssertForegroundThread();
+
+            if (_batchingProjectChanges)
+            {
+                // Already waiting to recompute host project, no need to do any more work to determine if we're dirty.
+                return;
```

Comment:
Yeah this kind of usage is a problem. You don't have a guarantee that the FG thread will see the 'false' value . You need a lock around that bool.

Reply:
That's ok though. We aren't passing around a stateful object. We re-compute the state at the time of `OnProjectChangedAsync`. So in the case that the background thread sets the `_batchingProjectChanges` to false (lets assume this if block has been entered) it'll already be queuing up changes based on the computed foreground state; otherwise this will have re-triggered another background computation. 

My intent here was for this to be a safe race. Maybe your race condition eyes are seeing something different though so feel free to call me out😄